Add a new host_identifier option

marpaia opened the pull request:
This PR does three basic things:

- Updates the number of `host_identifier` options to 3:
  - `uuid`: a universally unique identifier that is likely to be unique among all osquery instances
  - `hardware_id`: a identifier that it likely to be the same on instances of osquery running on the same hardware.
  - `hostname`: just the hostname of the computer. this is the default.
- Revert the functionality of `uuid` to that as it was before #2522 
- Add a `uuid` column to `osquery_info` which includes the osquery instance unique uuid (the `uuid` column of `system_info` will still return the hardware id).

osqueryer commented:
:-1: The commit 4f5597afd9ea6e8adcf760d4183bda534bcf506d (Job results: [114](https://jenkins.osquery.io/job/osqueryPullRequestOSX/114/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit 4f5597afd9ea6e8adcf760d4183bda534bcf506d (Job results: [3832](https://jenkins.osquery.io/job/osqueryPullRequestLinux/3832/)) failed one or more tests (**Linux**).

osqueryer commented:
:no_good: The commit 4f5597afd9ea6e8adcf760d4183bda534bcf506d (Job results: [814](https://jenkins.osquery.io/job/osqueryPullRequestAudit/814/)) failed the code audit or documentation test.

osqueryer commented:
:no_good: The commit 5e36c1065917587440a35c37ac9fa6808723ccab (Job results: [815](https://jenkins.osquery.io/job/osqueryPullRequestAudit/815/)) failed the code audit or documentation test.

theopolis commented:
How about adding a new UUID to `osquery_info` called an `instance_uuid` or `instance_id` and having it expose those information as well.

nit, perhaps `ephemeral` is a better term for the instance-specific UUID.

osqueryer commented:
:no_good: The commit 0ef75bec2cf99cb6118f71a76f1d7b9a44fc4639 (Job results: [816](https://jenkins.osquery.io/job/osqueryPullRequestAudit/816/)) failed the code audit or documentation test.

marpaia commented:
@theopolis i updated the PR to also persist the generated uuid to rocksdb , so it's a little less ephemeral, so I updated the name to `generated`, but I'd be happy to change it to whatever works.

osqueryer commented:
:-1: The commit 8561b42699dbddb6310a72b7c31b449f4c885868 (Job results: [118](https://jenkins.osquery.io/job/osqueryPullRequestOSX/118/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit 8561b42699dbddb6310a72b7c31b449f4c885868 (Job results: [3836](https://jenkins.osquery.io/job/osqueryPullRequestLinux/3836/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 8561b42699dbddb6310a72b7c31b449f4c885868 (Job results: [834](https://jenkins.osquery.io/job/osqueryPullRequestWindows/834/)) failed one or more tests (**Windows**).

marpaia commented:
I updated the options added by this PR and I updated the description of this PR to include a summary. I think it makes more sense to have the "generated uuid" from boost be `uuid` because it is a Universally Unique Identifier, whereas the implementation I re-named to `hardware_id` is more of just that- an identifier that is closely tied to the hardware that osquery is running on.

theopolis commented:
That makes sense.

I think this code needs a bit more organization, and the multiple `static`s are worrying me. But that cleanup can happen in another diff.

Just want to make sure folks are OK with the ambiguity of the `uuid` field. I think it works, `osquery_info` vs. `system_info` is grokable.

theopolis commented:
This is ready to go as soon as a CLA is completed.

fredemmott commented:
* bot poke *

theopolis commented:
Ping @marpaia, are there plans to implement the approach from the linked discussion issue in this PR?

marpaia commented:
Hey @theopolis, yeah, I'm sorry, I've been swamped with work. I'll see if I can get this done this week.

marpaia commented:
@theopolis just reality checking on this towards the end of the week. I'm probably not going to get to this by the end of tomorrow, but it's definitely on my short list of things to work on!

marpaia commented:
closing in favor of #2944
